app/src/main/java/com/hedvig/app/feature/genericauth/otpinput/SendOtpCodeUseCase.kt

```diff
@@ -0,0 +1,54 @@
+package com.hedvig.app.feature.genericauth.otpinput
+
+import androidx.annotation.StringRes
+import com.apollographql.apollo.ApolloClient
+import com.hedvig.android.owldroid.graphql.SendOtpCodeMutation
+import com.hedvig.app.R
+import com.hedvig.app.util.ErrorEvent
+import com.hedvig.app.util.apollo.QueryResult
+import com.hedvig.app.util.apollo.safeQuery
+
+class SendOtpCodeUseCase(
+    private val apolloClient: ApolloClient
+) {
+    suspend operator fun invoke(otpId: String, otpCode: String): OtpResult {
+        return when (val result = apolloClient.mutate(SendOtpCodeMutation(otpId, otpCode)).safeQuery()) {
+            is QueryResult.Error -> OtpResult.NetworkError(result.message)
+            is QueryResult.Success -> parseSuccessResponse(result)
+        }
+    }
+
+    private fun parseSuccessResponse(result: QueryResult.Success<SendOtpCodeMutation.Data>): OtpResult {
+        val attempt = result.data.login_verifyOtpAttempt
+        return attempt.asVerifyOtpLoginAttemptError?.let {
+            val errorEvent = parseErrorCode(it.errorCode)
+            OtpResult.OtpError(errorEvent)
+        } ?: attempt.asVerifyOtpLoginAttemptSuccess?.let {
+            OtpResult.Success(it.accessToken)
+        } ?: OtpResult.OtpError(OtpErrorEvent.UNKNOWN)
+    }
+
+    private fun parseErrorCode(errorCode: String) = when (errorCode) {
+        "TOO_MANY_ATTEMPTS" -> OtpErrorEvent.TOO_MANY_ATTEMPTS
+        "ALREADY_COMPLETED" -> OtpErrorEvent.ALREADY_COMPLETED
+        "EXPIRED" -> OtpErrorEvent.EXPIRED
+        "WRONG_OTP" -> OtpErrorEvent.WRONG_OTP
+        else -> OtpErrorEvent.UNKNOWN
+    }
+
+    sealed class OtpResult {
+        data class Success(val authToken: String) : OtpResult()
+        data class NetworkError(val message: String?) : OtpResult()
+        data class OtpError(val error: ErrorEvent) : OtpResult()
+    }
+
+    enum class OtpErrorEvent(@StringRes private val resourceId: Int) : ErrorEvent {
```

Comment:
I feel like the view-layer might be leaking out a bit when string-resources find themselves down in use-cases 🤔

Reply:
Yeah I agree.. Do you know of any good solutions for error handling?

Reply:
It depends a bit on the use-case.

For attaching diagnostic information for developer purposes, I think having something like `interface Error { val message: String }` could be ok, but for more result-style types I prefer having the types be specified as members of a sealed class, and then matched upon later on.

It is a style of error handling that can be slightly verbose at times, but I find the trade-off worth it (no 'invisible' error-handling, like with exceptions, which often can turn into simply not handling errors).